### PR TITLE
Deprecate Lambdas as task actions

### LINKED ISSUE #10751: Deprecate using Java lambda as task action
We already turn off caching for task actions that are specified as lambdas. We should consider deprecating support completely.

----
/cc @gradle/build-cache

### PR DESCRIPTION
**Pull request by wolfs**
The implementation of task actions implemented via Java Lambdas can't be tracked. Let's deprecate using them.

Fixes #10751

### CODE REVIEW

**Comment by wolfs** [comment]
@bot-gradle test QFL

**Comment by bamboo** [comment]
I'd be curious to understand a little more why actions implemented by lambdas can't be tracked. Is that described somewhere?

**Comment by wolfs** [comment]
For tracking an implementation, we use the class loader hash and the class name. For a Lambda, the class name is not stable across JVMs, so we can't know if it changed it not. Or am I missing something here?

**Comment by wolfs** [comment]
@bamboo Here is the original issue with some more context: https://github.com/gradle/gradle/issues/5510
There is this test which shows the problem: https://github.com/gradle/gradle/blob/9f00a2888dd490bb56dc9375697455d95106ce72/subprojects/core/src/integTest/groovy/org/gradle/api/tasks/LambdaInputsIntegrationTest.groovy#L74-L108

**Comment by bamboo** [comment]
Thanks for the pointers, @wolfs. I'm wondering whether some of the things we're doing for the configuration cache could help improving the situation here. I'll report back if I discover something.

**Comment by wolfs** [comment]
@bot-gradle test this

**Comment by wolfs** [comment]
I am wondering if we should also remove the non-cacheable reasons for non-cacheable implementations and inputs, given that we catch those problems now as part of validation.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/LambdaInputsIntegrationTest.groovy [hunk 1]
```diff
@@ -89,49 +94,23 @@ class LambdaInputsIntegrationTest extends AbstractIntegrationSpec implements Dir
         buildFile.makeOlder()
 
         when:
+        expectThatExecutionOptimizationDisabledWarningIsDisplayed(executer, implementationUnknown { nestedProperty('action').implementedByLambda('LambdaActionOriginal').includeLink() })
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/LambdaInputsIntegrationTest.groovy:97]: Let's split this into lines.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/NestedInputIntegrationTest.groovy [hunk 2]
```diff
@@ -778,32 +777,34 @@ class NestedInputIntegrationTest extends AbstractIntegrationSpec implements Dire
...
-            "  Non-cacheable inputs: property 'bean' was loaded with an unknown classloader (class 'NestedBean')."
-    }
-
+    @ValidationTestFor(
+        ValidationProblemId.UNKNOWN_IMPLEMENTATION
+    )
     @ToBeFixedForConfigurationCache(because = "uses custom GroovyClassLoader")
-    def "task with nested bean loaded with custom classloader is never up-to-date"() {
+    def "task with nested bean loaded with custom classloader disables execution optimizations"() {
+        expectUnindentedValidationMessage()
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/NestedInputIntegrationTest.groovy:785]: Let's move this to `expectThatExecutionOptimizationDisabledWarningIsDisplayed()`.

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc [hunk 3]
```diff
@@ -111,6 +111,13 @@ For example, a dependency substitution rule `substitute(project(':a')).with(proj
 With chaining you can, for example, add a reason for a substitution like this:
 `substitute(project(':a')).using(project(':b')).because("a reason")`.
 
+[[java_lamdba_action]]
+=== Using Java lambdas as task actions
+
+When using a Java lambda to implement a task action, Gradle cannot track the implementation and the task will never be up-to-date of from the build cache.
+Since it is easy to add such a task action, using task actions implemented by Java lambdas is now deprecated.
+See <<validation_problems.adoc#implementation_unknown,Validation problems>> for more details how to fix the issue.
```
Comment (lptr) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc:119]: We should address Kotlin 1.5 lambdas here. How about Groovy 3 (and 4) lambdas/closures?

File: subprojects/docs/src/docs/userguide/troubleshooting/validation_problems.adoc [hunk 4]
```diff
@@ -360,3 +360,43 @@ To fix this problem, you have two options:
...
+
+From Java code, it is quite easy to accidentally use a Lambda to implement a task action.
+
+```java
+tasks.register("myTask",
+    myTask -> myTask.doLast(t -> System.out.println("Hello from task"))
+);
+```
+
+Gradle currently isn't able to track the implementation of this task action.
```
Comment (lptr) [inline subprojects/docs/src/docs/userguide/troubleshooting/validation_problems.adoc:387]:
```suggestion
Gradle isn't able to track the implementation of this task action.
```

File: subprojects/docs/src/docs/userguide/troubleshooting/validation_problems.adoc [hunk 5]
```diff
@@ -360,3 +360,43 @@ To fix this problem, you have two options:
...
+    myTask -> myTask.doLast(new Action<Task>() {
+        @Override
+        public void execute(Task task) {
+            System.out.println("Hello from task");
+        }
+    })
+);
+```
+
+For the second case, instead of loading your class by a classloader unknown to Gradle, you can use Gradle's built-in ways to load the class.
```
Comment (lptr) [inline subprojects/docs/src/docs/userguide/troubleshooting/validation_problems.adoc:401]:
```suggestion
For the case where Gradle cannot track the implementation because it was loaded by a classloader unknown to Gradle, you can use Gradle's built-in ways to load the class.
```

File: subprojects/execution/src/main/java/org/gradle/internal/execution/steps/ValidateStep.java [hunk 6]
```diff
@@ -70,6 +75,7 @@ public ValidateStep(
     public R execute(UnitOfWork work, BeforeExecutionContext context) {
         WorkValidationContext validationContext = context.getValidationContext();
         work.validate(validationContext);
+        context.getBeforeExecutionState().ifPresent(beforeExecutionState -> validateImplementations(work, beforeExecutionState, validationContext));
```
Comment (lptr) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/steps/ValidateStep.java:78]: Split this to two lines.
```suggestion
        context.getBeforeExecutionState()
            .ifPresent(beforeExecutionState -> validateImplementations(work, beforeExecutionState, validationContext));
```

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java [hunk 7]
```diff
@@ -167,14 +167,18 @@ private String normalize(LogContent output) {
...
-                result.add(line);
+                result.add(normalizeLambdaIds(line));
                 i++;
             }
         }
 
         return LogContent.of(result).withNormalizedEol();
     }
 
+    private String normalizeLambdaIds(String line) {
```
Comment (lptr) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java:178]: Add some Javadoc explaining why we need this.

File: subprojects/model-core/src/test/groovy/org/gradle/internal/reflect/validation/ValidationMessageCheckerTest.groovy [hunk 8]
```diff
@@ -770,6 +770,72 @@ Please refer to https://docs.gradle.org/current/userguide/validation_problems.ht
...
+            nestedProperty('action')
+            implementedByLambda('LambdaAction')
+            includeLink()
+        }
+
+        then:
+        outputEquals """
+Property 'action' was implemented by the Java lambda 'LambdaAction\$\$Lambda\$<non-deterministic>'. Using Java lambdas is not supported, use an (anonymous) inner class instead.
+
+Reason: Gradle cannot track inputs when it doesn't know their implementation.
```
Comment (lptr) [inline subprojects/model-core/src/test/groovy/org/gradle/internal/reflect/validation/ValidationMessageCheckerTest.groovy:788]:
```suggestion
Reason: Gradle cannot track inputs when it cannot uniquely identify their implementation.
```

File: subprojects/model-core/src/test/groovy/org/gradle/internal/reflect/validation/ValidationMessageCheckerTest.groovy [hunk 9]
```diff
@@ -770,6 +770,72 @@ Please refer to https://docs.gradle.org/current/userguide/validation_problems.ht
...
+
+Possible solution: Use an (anonymous) inner class instead.
```
Comment (lptr) [inline subprojects/model-core/src/test/groovy/org/gradle/internal/reflect/validation/ValidationMessageCheckerTest.groovy:790]:
```suggestion
Possible solution: Use an (anonymous inner) class instead.
```

File: subprojects/model-core/src/test/groovy/org/gradle/internal/reflect/validation/ValidationMessageCheckerTest.groovy [hunk 10]
```diff
@@ -770,6 +770,72 @@ Please refer to https://docs.gradle.org/current/userguide/validation_problems.ht
...
+        when:
+        render implementationUnknown(true) {
+            nestedProperty('action')
+            implementedByLambda('LambdaAction')
+            includeLink()
+        }
+
+        then:
+        outputEquals """
+Property 'action' was implemented by the Java lambda 'LambdaAction\$\$Lambda\$<non-deterministic>'. Using Java lambdas is not supported, use an (anonymous) inner class instead.
```
Comment (lptr) [inline subprojects/model-core/src/test/groovy/org/gradle/internal/reflect/validation/ValidationMessageCheckerTest.groovy:786]: Solution is redundant here.

**Comment by bamboo** [comment]
How would you feel about introducing a combinator to still allow actions as lambdas?

```
doLast(lambda(t -> System.out.println("Look, Ma! No classes!")))
```

**Comment by wolfs** [comment]
> How would you feel about introducing a combinator to still allow actions as lambdas?

It would be better if this would work out of the box. I suppose the combinator would use a serializable `Action` type as argument?

**Comment by bamboo** [comment]
> I suppose the combinator would use a serializable Action type as argument?

Exactly. We have been using this one internally,
https://github.com/gradle/gradle/blob/f16adecaa3222848f877caceb45e68c062a86a00/subprojects/core/src/main/java/org/gradle/api/internal/lambdas/SerializableLambdas.java#L36

**Comment by bamboo** [comment]
> Can't we automatically do this at compile time or via bytecode manipulation?

**Comment by bamboo** [comment]
I'm sorry, I'm wearing a pressure glove and ended up pressing buttons while moving my hand :)

**Comment by wolfs** [comment]
It seems like we currently don't even support serializable lambdas as task inputs. We would need to do some additional work to find out what to track there, since the class name is still changing even for those. We could probably do something after serializing the lambda, though I am not sure if it is worth the effort.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/LambdaInputsIntegrationTest.groovy (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/LambdaInputsIntegrationTest.groovy]: Done.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/NestedInputIntegrationTest.groovy (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/NestedInputIntegrationTest.groovy]: This call is not necessary any more, I removed the method.

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc (hunk 3 above)
Comment (wolfs) [reply subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc]: I added something about Groovy and Kotlin to the user manual.

File: subprojects/docs/src/docs/userguide/troubleshooting/validation_problems.adoc (hunk 4 above)
Comment (wolfs) [reply subprojects/docs/src/docs/userguide/troubleshooting/validation_problems.adoc]: Done.

File: subprojects/docs/src/docs/userguide/troubleshooting/validation_problems.adoc (hunk 5 above)
Comment (wolfs) [reply subprojects/docs/src/docs/userguide/troubleshooting/validation_problems.adoc]: Done.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/steps/ValidateStep.java (hunk 6 above)
Comment (wolfs) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/steps/ValidateStep.java]: Done.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java (hunk 7 above)
Comment (wolfs) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/OutputScrapingExecutionResult.java:184]: Done.

File: subprojects/model-core/src/test/groovy/org/gradle/internal/reflect/validation/ValidationMessageCheckerTest.groovy (hunk 8 above)
Comment (wolfs) [reply subprojects/model-core/src/test/groovy/org/gradle/internal/reflect/validation/ValidationMessageCheckerTest.groovy]: The reason is now dependent on the kind of unknown implementation, please take another look.

File: subprojects/model-core/src/test/groovy/org/gradle/internal/reflect/validation/ValidationMessageCheckerTest.groovy (hunk 9 above)
Comment (wolfs) [reply subprojects/model-core/src/test/groovy/org/gradle/internal/reflect/validation/ValidationMessageCheckerTest.groovy]: Done.

File: subprojects/model-core/src/test/groovy/org/gradle/internal/reflect/validation/ValidationMessageCheckerTest.groovy (hunk 10 above)
Comment (wolfs) [reply subprojects/model-core/src/test/groovy/org/gradle/internal/reflect/validation/ValidationMessageCheckerTest.groovy]: I removed the redundant solution.

**Comment by wolfs** [comment]
@bot-gradle test this

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/SnapshotTaskInputsBuildOperationResult.java [hunk 11]
```diff
@@ -213,19 +213,7 @@ public void accept(CachingInputs inputs) {
...
-                public Set<String> apply(CachingInputs inputs) {
-                    ImmutableSortedSet<String> invalidInputProperties = inputs.getNonCacheableInputProperties();
-                    if (invalidInputProperties.isEmpty()) {
-                        return null;
-                    }
-                    return invalidInputProperties;
-                }
-            })
-            .orElse(null);
+        return null;
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/SnapshotTaskInputsBuildOperationResult.java:216]: Is this always `null` now? If so, we don't need a method for this.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java [hunk 12]
```diff
@@ -58,10 +54,7 @@ public final void withImplementation(ImplementationSnapshot implementation) {
     @OverridingMethodsMustInvokeSuper
     protected void processImplementation(ImplementationSnapshot implementation) {
         if (implementation.isUnknown()) {
-            noCachingReasonsBuilder.add(new CachingDisabledReason(
-                NON_CACHEABLE_IMPLEMENTATION,
-                "Implementation type " + implementation.getUnknownReason()
-            ));
+            throw new GradleException("Cannot calculate build cache key for work item with unknown implementation");
```
Comment (lptr) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java:57]: Is there a more specific exception we could/should throw here?

Where does this message appear? I don't see it in any of the tests.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/ImplementationChanges.java [hunk 13]
```diff
@@ -50,42 +49,34 @@ public boolean accept(ChangeVisitor visitor) {
                 executable.getDisplayName(), previousImplementation.getTypeName(), currentImplementation.getTypeName()));
         }
         if (currentImplementation.isUnknown()) {
-            return visitor.visitChange(new DescriptiveChange("The type of %s %s",
-                    executable.getDisplayName(), currentImplementation.getUnknownReason()));
+            // We already validate that the current implementation can't be unknown.
+            throw new GradleException("Cannot determine changes for work with unknown implementation");
```
Comment (lptr) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/ImplementationChanges.java:53]: Is there a more specific exception to throw here?

Where does the message end up? I don't see it in tests.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/ImplementationChanges.java [hunk 14]
```diff
@@ -50,42 +49,34 @@ public boolean accept(ChangeVisitor visitor) {
                 executable.getDisplayName(), previousImplementation.getTypeName(), currentImplementation.getTypeName()));
         }
         if (currentImplementation.isUnknown()) {
```
Comment (lptr) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/ImplementationChanges.java:51]: Does it make sense to keep track of an unknown implementation anymore if all we do is throw an exception down the line?

File: subprojects/hashing/src/main/java/org/gradle/internal/hash/Hasher.java [hunk 15]
```diff
@@ -77,22 +77,6 @@
...
-    /**
-     * Whether the build cache hash is valid.
-     */
-    boolean isValid();
-
-    /**
-     * Reason why the hash is not valid.
-     */
-    String getInvalidReason();
-
```
Comment (lptr) [inline subprojects/hashing/src/main/java/org/gradle/internal/hash/Hasher.java:95]: This is good!

File: subprojects/model-core/src/main/java/org/gradle/internal/reflect/validation/TypeValidationProblemRenderer.java [hunk 16]
```diff
@@ -76,7 +76,7 @@ public static String renderMinimalInformationAbout(TypeValidationProblem problem
      * a better way to display multiline deprecation warnings
      */
     public static String convertToSingleLine(String message) {
-        return message.replaceAll("(\\r?\\n)+", ". ")
+        return message.replaceAll("(\\r?\\n *)+", ". ")
```
Comment (lptr) [inline subprojects/model-core/src/main/java/org/gradle/internal/reflect/validation/TypeValidationProblemRenderer.java:79]: Why is this change needed? What about `\t`? Can we use `\s*` instead?

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/SnapshotTaskInputsBuildOperationResult.java (hunk 11 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/SnapshotTaskInputsBuildOperationResult.java:216]: This method is called by the scan plugin to capture the properties loaded by an unknown classloader. So we still need the method, though we should talk to the GE team after merging this what changes they want/need to do on their side to accommodate the new behavior.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java (hunk 12 above)
Comment (wolfs) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java:57]: This exception never should be thrown, since validation fails in the case and then we won't calculate the build cache key. Shall we remove these checks altogether?

File: subprojects/model-core/src/main/java/org/gradle/internal/reflect/validation/TypeValidationProblemRenderer.java (hunk 16 above)
Comment (wolfs) [reply subprojects/model-core/src/main/java/org/gradle/internal/reflect/validation/TypeValidationProblemRenderer.java:79]: For the assertions, we convert the multiline error to the single line deprecation message. This change allows a multiline _indented_ error to be converted correctly to the single line deprecation message. This allowed me to get rid of the method which removed the indentation.

I wouldn't add any more regex logic here as long as the joining work and remove this method as soon as we have multiline deprecation messages.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java (hunk 12 above)
Comment (lptr) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java:57]: If it's an assertion, then we should throw `AssertionError`. But it would be a lot better to avoid this problem completely. I wonder if we can remove tracking unknown implementations. If not, then just not checking here is also a good alternative.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/SnapshotTaskInputsBuildOperationResult.java (hunk 11 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/SnapshotTaskInputsBuildOperationResult.java:216]: Ah, right. Got confused by this a bit, thinking it was one of the types we owned exclusively. Shall we deprecate the method?

File: subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/ImplementationChanges.java (hunk 14 above)
Comment (lptr) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/ImplementationChanges.java:51]: If we need the fact that the implementation is unknown, we could use an `Optional` to keep track of that, too. But my hunch is that we can get rid of tracking unknown completely.

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/SnapshotTaskInputsBuildOperationResult.java (hunk 11 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/SnapshotTaskInputsBuildOperationResult.java]: I deprecated the method.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java (hunk 12 above)
Comment (wolfs) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java]: I think we can't yet remove tracking unknown implementations, given that we still need to store them in the task history. As soon as we fail when we encounter an unknown implementation, I think we can stop tracking them.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/ImplementationChanges.java (hunk 13 above)
Comment (wolfs) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/ImplementationChanges.java]: I removed the throwing of the exception and a longer comment about what is going on at the top of the method.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java (hunk 12 above)
Comment (lptr) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java]: Since we don't track the individual reasons now for the implementation to be unknown, can we still ditch `ImplementationSnapshot.isUnknown()` and just pass around `null` or something?

Comment (wolfs) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java]: We need the reason for the validation, since this is what we'll print out. So we can't just use `null` as implementation snapshot.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java [hunk 17]
```diff
@@ -68,9 +64,6 @@ public final void withAdditionalImplementations(Iterable<ImplementationSnapshot>
 
     @OverridingMethodsMustInvokeSuper
     protected void processAdditionalImplementation(ImplementationSnapshot additionalImplementation) {
-        if (additionalImplementation.isUnknown()) {
-            throw new GradleException("Cannot calculate build cache key for additional action with unknown implementation");
-        }
     }
```
Comment (lptr) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java:67]: Nit: I wonder if these two empty implementations can be further simplified. We surely don't need the `@OverridingMethodsMustInvokeSuper` anymore.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/ImplementationChanges.java [hunk 18]
```diff
@@ -44,14 +43,13 @@ public ImplementationChanges(
 
     @Override
     public boolean accept(ChangeVisitor visitor) {
+        // In this method, the current implementation and additional implementations can't be unknown, since that causes already a validation warning.
+        // Previous implementations can still be unknown, since we store the inputs in the task history even if validation fails.
+        // When we fail the build for unknown implementations, then the previous implementations also can't be unknown.
```
Comment (lptr) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/ImplementationChanges.java:48]: This would be so much better to capture by using the type system, say, by introducing something like `PotentiallyUnknownImplementationSnapshot`, or something.

Comment (wolfs) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/history/changes/ImplementationChanges.java]: I did a change in that direction.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java (hunk 17 above)
Comment (wolfs) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/caching/impl/DefaultCachingStateBuilder.java:65]: I removed this annotation.

**Comment by wolfs** [comment]
@bot-gradle test this

**Review by lptr — APPROVED** [review]
LGTM. I'd love to not expose `KnownImplementationSnapshot` but instead use something like `Either<ImplementationSnapshot, String>` to carry the reason when implementation is not known.

**Comment by wolfs** [comment]
@bot-gradle test and merge
